### PR TITLE
Fix missing assemblyref for typeref only kept for debug info in illink

### LINKED ISSUE #86462: ILLink produces invalid IL metadata with -b switch and Save assembly mode
As part of testing https://github.com/xamarin/xamarin-macios/pull/18268 we hit an odd case where invalid IL is produced as the output of ILLink, resulting in later crash when post-processing the data with ILStrip. There's actually several issues at play here:
- Roslyn produces an unused TypeRef for global using aliases (https://github.com/dotnet/roslyn/issues/68250). Xamarin uses this to create an alias for `nfloat`.
- ILLink with `-b` switch strips the unused AssemblyRef but keeps the unused TypeRef in the `Save` assembly mode. Xamarin uses ILLink, among other things, to produce the Managed Static Registrar. In this configuration the original `Copy` mode of the input assembly is changed to `Save` in the process.
- ILStrip crashes on the invalid input (Xamarin uses pretty old version, bump is handled in https://github.com/xamarin/xamarin-macios/pull/18291).

Repro:
- Download [extra-typeref-linker.zip](https://github.com/dotnet/runtime/files/11510821/extra-typeref-linker.zip)
- Update the illinktask.dll path in linker.rsp to an absolute path to the included task .dll; the custom task only changes the assembly action to `Save` since it cannot be done on command line
- Run `dotnet <path to illink.dll> @linker.rsp`
- Check `out\extra-typeref.dll` with the `mdv` tool (https://github.com/dotnet/metadata-tools) and observe that there's `NFloat` type reference that points to non-existent assembly reference

Observations:
- Tested with .NET 7.0.302 and .NET 8.0 Preview 5 linkers. The issue exists in both. On .NET 8 the original test case produces different linker input and doesn't hit the problem, but the underlying bug still exists.
- Removing the `-b` switch from `linker.rsp` makes the problem go away

cc @rolfbjarne @ivanpovazan

### PR DESCRIPTION
**Pull request by sbomer**
Fixes https://github.com/dotnet/runtime/issues/86462 using the approach suggested in https://github.com/dotnet/runtime/issues/86462#issuecomment-1554423173. Initially I was looking into removing the unused typeref, but since we don't sweep debug info for save assemblies, I think it makes sense to keep it, even though this means the presence of the typeref in the final output depends on whether we are linking the PDB.

The fix also needed to walk up the parent import scopes, discovered by reproducing the issue in our test infra.

The behavior needs to depend on whether we are linking debug symbols, otherwise we will keep the assemblyref (but not the typeref that uses it) when PDBs are present, and I don't think the output should depend on the presence of PDBs. `AssemblyOnlyUsedByUsingSaveAction` tests this case - see comments in there for more detail.

This includes some unrelated test infra changes (supporting multiple additional compiler arguments) which were useful for iterating on this, even though they aren't necessary in the final version of the testcases.

### CODE REVIEW

**Review by filipnavara — APPROVED** [review]
Thanks! I got stuck on making the unit test for this, so I am glad you tackled it in the end.

File: src/tools/illink/test/Mono.Linker.Tests/TestCasesRunner/ResultChecker.cs [hunk 1]
```diff
@@ -49,13 +49,43 @@ public ResultChecker ()
...
-			foreach (var result in verifier.Results) {
-				if (result.Code == ILVerify.VerifierError.None)
-					continue;
+			foreach (var result in verifier.Results)
 				Assert.Fail (ILVerifier.GetErrorMessage (result));
+		}
+
+		static void ValidateTypeRefsHaveValidAssemblyRefs (AssemblyDefinition linked)
+		{
+			Console.WriteLine ("Validating type references have valid assembly references: " + linked.FullName);
```
Comment (vitek-karas) [inline src/tools/illink/test/Mono.Linker.Tests/TestCasesRunner/ResultChecker.cs:61]: Where does this go?
Just curious - I haven't seen us using Console.WriteLine in tests for anything yet.

**Review by vitek-karas — APPROVED** [review]
Looks good - thanks.

I find it weird that we intentionally leave a type ref/assembly ref in but remove the target from the app - but I guess that's what `save` is about. I got too used to true trimming recently with AOT and such.

**Comment by vitek-karas** [comment]
I perfectly fine with the behavior regarding stripping symbols - I think it makes sense.

**Pull request merged**